### PR TITLE
Move to something better.

### PR DESCRIPTION
**Pull request by dinosaure**
Clearly, this PR need a big clean up. At the end, we use `bechamel` and lost the support of `4.03.0`, `4.04.0` and `4.05.0`. So, I'm not happy with this situation.

### CODE REVIEW

File: lib/eqaf.ml [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
+  for i = 0 to pred (len / 8) do
+    res :=
+      Int64.logor !res
+        (Int64.logxor
+           (unsafe_string_get_64 a (i * 8))
+           (unsafe_string_get_64 a (i * 8)))
+  done ;
+  for i = len - (len mod 8) to pred len do
+    res :=
+      Int64.logor !res
```
Comment (cfcs) [inline lib/eqaf.ml]: Why do you use Int64 instead eg Nativeint here? Won't the 64bit expansion be slower on 32bit?

Comment (dinosaure) [reply lib/eqaf.ml]: I did a benchmark for you to compare `eqnt` which use the `Native` module and `eqbg`/`eqaf` which uses `Int64` and I got this result on my server (no virtualization):

```sh
┌─────────────────┬─────────────┬────────────┐
│ Name            │    Time/Run │ Percentage │
├─────────────────┼─────────────┼────────────┤
│ diff eqaf:0     │     21.48ns │      0.16% │
│ diff eqaf:10    │     25.27ns │      0.18% │
│ diff eqaf:100   │     42.39ns │      0.31% │
│ diff eqaf:1000  │    186.02ns │      1.36% │
│ diff eqaf:2000  │    338.12ns │      2.47% │
│ diff eqaf:4000  │    642.03ns │      4.68% │
│ diff eqaf:8000  │  1_249.76ns │      9.12% │
│ diff eqaf:10000 │  1_553.53ns │     11.34% │
│ same eqaf:0     │     21.49ns │      0.16% │
│ same eqaf:10    │     25.25ns │      0.18% │
│ same eqaf:100   │     42.38ns │      0.31% │
│ same eqaf:1000  │    186.04ns │      1.36% │
│ same eqaf:2000  │    338.02ns │      2.47% │
│ same eqaf:4000  │    642.03ns │      4.68% │
│ same eqaf:8000  │  1_249.83ns │      9.12% │
│ same eqaf:10000 │  1_553.38ns │     11.33% │
│ diff eqst:0     │      7.75ns │      0.06% │
│ diff eqst:10    │      7.75ns │      0.06% │
│ diff eqst:100   │      7.75ns │      0.06% │
│ diff eqst:1000  │      7.75ns │      0.06% │
│ diff eqst:2000  │      7.76ns │      0.06% │
│ diff eqst:4000  │      7.76ns │      0.06% │
│ diff eqst:8000  │      7.75ns │      0.06% │
│ diff eqst:10000 │      7.75ns │      0.06% │
│ same eqst:0     │      7.75ns │      0.06% │
│ same eqst:10    │      8.66ns │      0.06% │
│ same eqst:100   │     25.84ns │      0.19% │
│ same eqst:1000  │     95.72ns │      0.70% │
│ same eqst:2000  │    156.74ns │      1.14% │
│ same eqst:4000  │    277.51ns │      2.02% │
│ same eqst:8000  │    565.09ns │      4.12% │
│ same eqst:10000 │    643.67ns │      4.70% │
│ diff eqnt:0     │     18.33ns │      0.13% │
│ diff eqnt:10    │     31.00ns │      0.23% │
│ diff eqnt:100   │    161.82ns │      1.18% │
│ diff eqnt:1000  │  1_392.18ns │     10.16% │
│ diff eqnt:2000  │  2_759.05ns │     20.13% │
│ diff eqnt:4000  │  5_496.20ns │     40.10% │
│ diff eqnt:8000  │ 10_963.76ns │     80.00% │
│ diff eqnt:10000 │ 13_693.26ns │     99.91% │
│ same eqnt:0     │     18.69ns │      0.14% │
│ same eqnt:10    │     30.99ns │      0.23% │
│ same eqnt:100   │    162.40ns │      1.18% │
│ same eqnt:1000  │  1_392.54ns │     10.16% │
│ same eqnt:2000  │  2_759.47ns │     20.13% │
│ same eqnt:4000  │  5_494.26ns │     40.09% │
│ same eqnt:8000  │ 10_962.34ns │     79.99% │
│ same eqnt:10000 │ 13_702.32ns │     99.98% │
│ diff eqbg:0     │     20.37ns │      0.15% │
│ diff eqbg:10    │     24.71ns │      0.18% │
│ diff eqbg:100   │     41.03ns │      0.30% │
│ diff eqbg:1000  │    183.84ns │      1.34% │
│ diff eqbg:2000  │    335.77ns │      2.45% │
│ diff eqbg:4000  │    639.62ns │      4.67% │
│ diff eqbg:8000  │  1_246.88ns │      9.10% │
│ diff eqbg:10000 │  1_551.20ns │     11.32% │
│ same eqbg:0     │     20.36ns │      0.15% │
│ same eqbg:10    │     24.71ns │      0.18% │
│ same eqbg:100   │     41.03ns │      0.30% │
│ same eqbg:1000  │    183.89ns │      1.34% │
│ same eqbg:2000  │    335.36ns │      2.45% │
│ same eqbg:4000  │    639.38ns │      4.67% │
│ same eqbg:8000  │  1_247.10ns │      9.10% │
│ same eqbg:10000 │  1_551.12ns │     11.32% │
│ diff eqml:0     │      5.47ns │      0.04% │
│ diff eqml:10    │     19.45ns │      0.14% │
│ diff eqml:100   │    151.40ns │      1.10% │
│ diff eqml:1000  │  1_382.76ns │     10.09% │
│ diff eqml:2000  │  2_752.33ns │     20.08% │
│ diff eqml:4000  │  5_490.57ns │     40.06% │
│ diff eqml:8000  │ 10_969.67ns │     80.04% │
│ diff eqml:10000 │ 13_701.77ns │     99.98% │
│ same eqml:0     │      5.93ns │      0.04% │
│ same eqml:10    │     19.45ns │      0.14% │
│ same eqml:100   │    152.04ns │      1.11% │
│ same eqml:1000  │  1_382.86ns │     10.09% │
│ same eqml:2000  │  2_751.57ns │     20.08% │
│ same eqml:4000  │  5_487.31ns │     40.04% │
│ same eqml:8000  │ 10_969.40ns │     80.04% │
│ same eqml:10000 │ 13_705.11ns │    100.00% │
└─────────────────┴─────────────┴────────────┘
```

As you can see `eqbg` is more fast than `eqnt`: `1551ns < 13702ns` for `string` with a length `= 10000`, so I think, we will keep `eqbg`/`eqaf` of this PR

**Comment by hannesm** [comment]
OCaml versions: why is this bad? For MirageOS, we agreed that the "most recent two major" are in our goal -- that soon means 4.07 and 4.08.

**Comment by dinosaure** [comment]
Ok I will clean this PR and merge it and make a release after 👍 !

**Comment by hannesm** [comment]
from a security perspective, since 4.06.0 has safe-string by default this is nice, but since there have been some runtime and soundness issues discovered in OCaml, always using the most recent version is a good idea anyways. obviously there may be performance regressions (that's why sticking to "most two recent major versions" is a good tradeoff imho)

File: lib/eqaf.ml (hunk 1 above)
Comment (cfcs) [reply lib/eqaf.ml]: Does your server run on a 32bit processor?

Comment (dinosaure) [reply lib/eqaf.ml]: `amd64`, I did not have an access to a 32-bits machine :(

File: lib/eqaf.ml [hunk 2]
```diff
@@ -1,11 +1,21 @@
-external unsafe_string_get : string -> int -> int = "%string_unsafe_get"
+external unsafe_string_get : string -> int -> int = "%string_unsafe_get" [@@noalloc]
+external unsafe_string_get_64 : string -> int -> int64 = "%caml_string_get64u"
```
[the remark this answers is not in the document]
Comment (cfcs) [reply lib/eqaf.ml]: Yes :-(

**Comment by dinosaure** [comment]
Ok, I decided to move `eqaf` to something else. `bechamel` is good enough to get and compare results but I don't want to have a (big) dependency on `eqaf` where this package provide only one function. So I took what is really interesting from `bechamel` and make my personal tool to check if `eqaf` has our assumptions (constant-time).

The idea behind `check` is to do a linear regression (ordinary least square) as `bechamel` did. However, instead to compare results from regression directly, I found a proper way to compare regression coefficients between two groups.

- The first group is time needed to compare two values which are the same.
- The second group is time needed to compare two values which are different.

So, we need to make a dummy variable that is coded `0` for the first group and `1` for the second group. Then, we need to make a new variable which will be the product between the dummy variable and how many run we did. Finally, we will use dummy variable, _run_ and the product as predictors in the regression equation.

Result for `eqaf` should be close to `0.0` where the first group (time needed to execute `eqaf`) has no difference with the second group. I ran this kind of test with `String.compare` too and result is close to `200.0` - and this is what we expect.

So, it's mostly work! However, virtual environment is hard and not really predictable. Indeed, in a virtual host system (like mine), we have some noises when we execute `equal` function which can impact a lot results from linear regression. I don't have any way to fix it and CI should fail sometimes ... From experimental feedback, it's mostly about Windows CI.

Finally, this branch add `equal_bigstring` and `equal_cstruct` as optional libraries. And then, core package does not need anything to provide library and test.

**Comment by dinosaure** [comment]
@hannesm @cfcs just to ask you a feedback about this PR :)

**Comment by hannesm** [comment]
sounds good to me, I won't have time to review this PR in detail..

**Pull request merged**